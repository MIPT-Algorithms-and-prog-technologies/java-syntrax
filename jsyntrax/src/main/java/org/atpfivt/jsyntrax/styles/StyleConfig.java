package org.atpfivt.jsyntrax.styles;

import org.atpfivt.jsyntrax.Main;
import org.atpfivt.jsyntrax.util.StringUtils;
import org.ini4j.Profile.Section;
import org.ini4j.Wini;

import java.awt.Color;
import java.awt.Font;
import java.io.IOException;
import java.lang.reflect.Field;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;
import java.util.regex.Pattern;


public final class StyleConfig {
    private int lineWidth = 2;
    private int outlineWidth = 2;
    private int padding = 5;
    private Color lineColor = new Color(0, 0, 0);
    private int maxRadius = 9;
    private int hSep = 17;
    private int vSep = 9;
    private boolean arrows = true;
    private TitlePosition titlePos = TitlePosition.tl;
    private Color bulletFill = new Color(255, 255, 255);
    private Color textColor = new Color(0, 0, 0);
    private boolean shadow = true;
    private Color shadowFill = new Color(0, 0, 0, 127);
    private Font titleFont = new Font("Sans", Font.BOLD, 22);
    private double scale;
    private boolean transparency;
    private NodeStyle defNodeStyle = new NodeStyle();
    private List<NodeStyle> nodeStyles = new ArrayList<>(List.of(
            new NodeBubbleStyle(),
            new NodeBoxStyle(),
            new NodeTokenStyle(),
            new NodeHexStyle()
    ));


    private void parseStyleArgs(Section s) throws IOException {
        try {
            parseField(s, this, "line_width", Integer::parseInt);
            parseField(s, this, "outline_width", Integer::parseInt);
            parseField(s, this, "padding", Integer::parseInt);
            parseField(s, this, "line_color", StringUtils::colorFromString);
            parseField(s, this, "max_radius", Integer::parseInt);
            parseField(s, this, "h_sep", Integer::parseInt);
            parseField(s, this, "v_sep", Integer::parseInt);
            parseField(s, this, "arrows", Boolean::parseBoolean);
            parseField(s, this, "title_pos", (String v) -> TitlePosition.valueOf(v.substring(1, v.length() - 1)));
            parseField(s, this, "bullet_fill", StringUtils::colorFromString);
            parseField(s, this, "text_color", StringUtils::colorFromString);
            parseField(s, this, "shadow", Boolean::parseBoolean);
            parseField(s, this, "shadow_fill", StringUtils::colorFromString);
            parseField(s, this, "title_font", StringUtils::fontFromString);
        } catch (NoSuchFieldException | IllegalAccessException e) {
            throw new IOException(e);
        }
    }

    private void parseNodeStyle(Section s, NodeStyle ns) throws IOException {
        try {
            Function<String, Object> patFunc = (String v) -> Pattern.compile(v.substring(1, v.length() - 1));
            parseField(s, ns, "pattern", patFunc);
            parseField(s, ns, "shape", (String v) -> v.substring(1, v.length() - 1));
            parseField(s, ns, "font", StringUtils::fontFromString);
            parseField(s, ns, "text_color", StringUtils::colorFromString);
            parseField(s, ns, "fill", StringUtils::colorFromString);
        } catch (NoSuchFieldException | IllegalAccessException e) {
            throw new IOException(e);
        }
    }


    private void parseField(Section s, Object obj,
                            String propertyName, Function<String, Object> parser)
            throws NoSuchFieldException, IllegalAccessException {
        if (s.containsKey(propertyName)) {
            Field field = obj.getClass()
                    .getDeclaredField(StringUtils.snakeToCamelCase(propertyName));
            field.setAccessible(true);
            field.set(obj, parser.apply(s.get(propertyName)));
        } else {
            System.out.println(
                    "[StyleConfigParser] Warning: section ["
                            + s.getName() + "]"
                            + " does not contain property ["
                            + propertyName + "]");
        }
    }


    public StyleConfig(double scale, boolean transparency) throws IOException {
        this.setTransparency(transparency);
        this.setScale(scale);

        Wini config = new Wini(Main.class
                .getResourceAsStream("/jsyntrax.ini")
        );
        if (config.containsKey("style")) {
            parseStyleArgs(config.get("style"));
        }
    }


    public StyleConfig(double scale, boolean transparency, Path style) throws IOException {
        this.setTransparency(transparency);
        this.setScale(scale);

        Wini ini = new Wini(style.toFile());
        if (ini.containsKey("style")) {
            parseStyleArgs(ini.get("style"));
        }
        getNodeStyles().clear();
        for (Section s : ini.values()) {
            if (s.getName().equals("style")) {
                continue;
            }
            // this is custom node style
            NodeStyle ns = new NodeStyle();
            ns.setName(s.getName());
            parseNodeStyle(s, ns);
            getNodeStyles().add(ns);
        }
    }


    public NodeStyle getNodeStyle(String txt) {
        for (NodeStyle ns : getNodeStyles()) {
            if (ns.match(txt)) {
                return ns;
            }
        }
        return getDefNodeStyle();
    }


    public boolean getArrows() {
        return isArrows();
    }

    public int getLineWidth() {
        return lineWidth;
    }

    public int getOutlineWidth() {
        return outlineWidth;
    }

    public int getPadding() {
        return padding;
    }

    public Color getLineColor() {
        return lineColor;
    }

    public int getMaxRadius() {
        return maxRadius;
    }

    public int getHSep() {
        return gethSep();
    }

    public int getVSep() {
        return getvSep();
    }

    public TitlePosition getTitlePos() {
        return titlePos;
    }

    public Color getBulletFill() {
        return bulletFill;
    }

    public Color getTextColor() {
        return textColor;
    }

    public boolean isShadow() {
        return shadow;
    }

    public Color getShadowFill() {
        return shadowFill;
    }

    public Font getTitleFont() {
        return titleFont;
    }

    public List<NodeStyle> getNodeStyles() {
        return nodeStyles;
    }

    public double getScale() {
        return scale;
    }

    public boolean isTransparent() {
        return isTransparency();
    }

    public int gethSep() {
        return hSep;
    }

    public int getvSep() {
        return vSep;
    }

    public boolean isArrows() {
        return arrows;
    }

    public boolean isTransparency() {
        return transparency;
    }

    public void setLineWidth(int lineWidth) {
        this.lineWidth = lineWidth;
    }

    public void setOutlineWidth(int outlineWidth) {
        this.outlineWidth = outlineWidth;
    }

    public void setPadding(int padding) {
        this.padding = padding;
    }

    public void setLineColor(Color lineColor) {
        this.lineColor = lineColor;
    }

    public void setMaxRadius(int maxRadius) {
        this.maxRadius = maxRadius;
    }

    public void sethSep(int hSep) {
        this.hSep = hSep;
    }

    public void setvSep(int vSep) {
        this.vSep = vSep;
    }

    public void setArrows(boolean arrows) {
        this.arrows = arrows;
    }

    public void setTitlePos(TitlePosition titlePos) {
        this.titlePos = titlePos;
    }

    public void setBulletFill(Color bulletFill) {
        this.bulletFill = bulletFill;
    }

    public void setTextColor(Color textColor) {
        this.textColor = textColor;
    }

    public void setShadow(boolean shadow) {
        this.shadow = shadow;
    }

    public void setShadowFill(Color shadowFill) {
        this.shadowFill = shadowFill;
    }

    public void setTitleFont(Font titleFont) {
        this.titleFont = titleFont;
    }

    public void setScale(double scale) {
        this.scale = scale;
    }

    public void setTransparency(boolean transparency) {
        this.transparency = transparency;
    }

    public NodeStyle getDefNodeStyle() {
        return defNodeStyle;
    }

    public void setDefNodeStyle(NodeStyle defNodeStyle) {
        this.defNodeStyle = defNodeStyle;
    }

    public void setNodeStyles(List<NodeStyle> nodeStyles) {
        this.nodeStyles = nodeStyles;
    }
}

