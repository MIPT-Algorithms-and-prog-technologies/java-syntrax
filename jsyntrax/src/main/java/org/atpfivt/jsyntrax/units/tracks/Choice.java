package org.atpfivt.jsyntrax.units.tracks;

import org.atpfivt.jsyntrax.units.Unit;
import org.atpfivt.jsyntrax.visitors.Visitor;

import java.util.ArrayList;

public class Choice extends ComplexTrack {
  public Choice(ArrayList<Unit> units) {
    super(units);
  }

  public void accept(Visitor visitor) {
    visitor.visitChoice(this);
  }
}
