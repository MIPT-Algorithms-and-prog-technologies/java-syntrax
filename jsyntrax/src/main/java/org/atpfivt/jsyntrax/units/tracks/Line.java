package org.atpfivt.jsyntrax.units.tracks;

import org.atpfivt.jsyntrax.units.Unit;
import org.atpfivt.jsyntrax.visitors.Visitor;

import java.util.ArrayList;

public class Line extends Track {
  public Line(ArrayList<Unit> units) {
    super(units);
  }

  public Line(Unit unit) {
    super(new ArrayList<>());
    units.add(unit);
  }

  public void accept(Visitor visitor) {
    visitor.visitLine(this);
  }
}
