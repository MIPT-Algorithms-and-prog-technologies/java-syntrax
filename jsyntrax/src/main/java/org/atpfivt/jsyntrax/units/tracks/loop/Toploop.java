package org.atpfivt.jsyntrax.units.tracks.loop;

import org.atpfivt.jsyntrax.exceptions.LoopNotTwoArgsException;
import org.atpfivt.jsyntrax.units.Unit;
import org.atpfivt.jsyntrax.visitors.Visitor;

import java.util.List;

public class Toploop extends Loop {
  public Toploop(List<Unit> units) throws LoopNotTwoArgsException {
    super(units);
  }

  public void accept(Visitor visitor) {
    visitor.visitToploop(this);
  }
}
