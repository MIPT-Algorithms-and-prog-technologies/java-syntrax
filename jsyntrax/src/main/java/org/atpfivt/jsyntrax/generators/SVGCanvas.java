package org.atpfivt.jsyntrax.generators;

import org.atpfivt.jsyntrax.generators.elements.Element;
import org.atpfivt.jsyntrax.styles.NodeStyle;
import org.atpfivt.jsyntrax.styles.Style;
import org.atpfivt.jsyntrax.util.Algorithm;
import org.atpfivt.jsyntrax.util.Pair;

import java.awt.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public class SVGCanvas {

    public SVGCanvas(Style style, Map<String, String> urlMap) {
        this.urlMap = urlMap;
        this.style = style;
        this.tagcnt = new HashMap<>();
        this.elements = new ArrayList<>();
    }

    // default prefix = "x", default suffix = ""
    public String new_tag(String prefix, String suffix) {
        String f = prefix + "___" + suffix;
        Integer value = tagcnt.getOrDefault(f, 0);
        tagcnt.put(f, value + 1);
        return prefix + value.toString() + suffix;
    }

    public void addElement(Element e) {
        this.elements.add(e);
    }

    public void addTagByTag(String addTag, String tag) {
        for (Element e : elements) {
            if (e.isTagged(tag)) {
                e.addTag(addTag);
            }
        }
    }

    public void dropTag(final String tag) {
        for (Element e : elements) {
            if (e.isTagged(tag)) {
                e.delTag(tag);
            }
        }
    }

    public void moveByTag(String tag, int dx, int dy) {
        for (Element e : elements) {
            if (e.isTagged(tag)) {
                e.start.f += dx;
                e.start.s += dy;
                e.end.f += dx;
                e.end.s += dy;
            }
        }
    }

    public void scaleByTag(String tag, double scale) {
        for (Element e : elements) {
            if (e.isTagged(tag)) {
                e.scale(scale);
            }
        }
    }

    public Optional<String> getCanvasTag() {
        return elements.stream().findAny().flatMap(Element::getAnyTag);
    }

    public Pair<Pair<Integer, Integer>, Pair<Integer, Integer>> getBoundingBoxByTag(String tag) {
        Pair<Integer, Integer> start = null;
        Pair<Integer, Integer> end = null;
        for (Element e: this.elements)
        {
            if (e.isTagged(tag)) {
                if (start == null) {
                    start = new Pair<>(e.start);
                    end = new Pair<>(e.end);
                }
                start.f = Math.min(start.f, e.start.f);
                start.f = Math.min(start.f, e.end.f);
                start.s = Math.min(start.s, e.start.s);
                start.s = Math.min(start.s, e.end.s);

                end.f = Math.max(end.f, e.start.f);
                end.f = Math.max(end.f, e.end.f);
                end.s = Math.max(end.s, e.start.s);
                end.s = Math.max(end.s, e.end.s);
            }
        }
        return new Pair<>(start, end);
    }

    public String generateSVG() {
        StringBuilder sb = new StringBuilder();
        double scale = style.scale;

        Pair<Pair<Integer, Integer>, Pair<Integer, Integer>> res = getBoundingBoxByTag("all");

        // move to picture to (0, 0)
        moveByTag("all", -res.f.f, -res.f.s);
        scaleByTag("all", scale);
        moveByTag("all", style.padding, style.padding);

        res = getBoundingBoxByTag("all");
        Pair<Integer, Integer> end = res.s;

        int W = end.f + style.padding;
        int H = end.s + style.padding;

        // collect fonts
        HashMap<String, Pair<Font, Color>> fonts = new HashMap<>();
        fonts.put("title_font", new Pair<>(this.style.title_font, this.style.text_color));
        for (NodeStyle ns : this.style.nodeStyles) {
            fonts.put(ns.name + "_font", new Pair<>(ns.font, ns.text_color));
        }

        // header
        sb.append("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n")
                .append("<!-- Created by Jyntrax https://github.com/atp-mipt/jsyntrax -->\n");
        sb.append("<svg xmlns=\"http://www.w3.org/2000/svg\"\n");
        sb.append("xmlns:xlink=\"http://www.w3.org/1999/xlink\"\n");
        sb.append("xml:space=\"preserve\"\n");
        sb.append("width=\"").append(W).append("\" ")
                .append("height=\"").append(H).append("\" ")
                .append("version=\"1.1\">\n");
        // styles
        sb.append("<style type=\"text/css\">\n");
        sb.append("<![CDATA[\n");
        // fonts
        for (Map.Entry<String, Pair<Font, Color>> fontPair : fonts.entrySet()) {
            String fontName = fontPair.getKey();
            String fontFamily = fontPair.getValue().f.getName();
            String fontSize = Integer.toString((int)(fontPair.getValue().f.getSize() * scale));
            String fontWeight = "normal";
            if ((fontPair.getValue().f.getStyle() & Font.BOLD) == Font.BOLD) {
                fontWeight = "bold";
            }
            String fontStyle = "normal";
            if ((fontPair.getValue().f.getStyle() & Font.ITALIC) == Font.ITALIC) {
                fontStyle = "italic";
            }

            String hex = Algorithm.toHex(fontPair.getValue().s);

            sb.append(".").append(fontName).append(" ");
            sb.append("{fill:").append(hex).append("; text-anchor:middle;\n");
            sb.append("font-family:").append(fontFamily).append("; ");
            sb.append("font-size:").append(fontSize).append("pt; ");
            sb.append("font-weight:").append(fontWeight).append("; ");
            sb.append("font-style:").append(fontStyle).append("; ");
            sb.append("}\n");
        }
        // other fonts
        sb.append(".label {fill: #000; text-anchor:middle; font-size:16pt; font-weight:bold; font-family:Sans;}\n");
        sb.append(".link {fill: #0D47A1;}\n");
        sb.append(".link:hover {fill: #0D47A1; text-decoration:underline;}\n");
        sb.append(".link:visited {fill: #4A148C;}\n");
        // close
        sb.append("]]>\n</style>\n");
        // defs
        sb.append("<defs>\n");
        sb.append("<marker id=\"arrow\" markerWidth=\"5\" markerHeight=\"4\" ")
                .append("refX=\"2.5\" refY=\"2\" orient=\"auto\" markerUnits=\"strokeWidth\">\n");
        String hex = Algorithm.toHex(this.style.line_color);
        sb.append("<path d=\"M0,0 L0.5,2 L0,4 L4.5,2 z\" fill=\"").append(hex).append("\" />\n");
        sb.append("</marker>\n</defs>\n");

        // elements
        if (!style.transparent) {
            sb.append("<rect width=\"100%\" height=\"100%\" fill=\"white\"/>\n");
        }
        for (Element e : this.elements) {

            if (style.shadow) {
                e.addShadow(sb, this.style);
            }
            e.toSVG(sb, this.style);
        }
        // end
        sb.append("</svg>\n");
        return sb.toString();
    }

    public final Map<String, String> urlMap;

    private final Style style;

    private final HashMap<String, Integer> tagcnt;

    private final ArrayList<Element> elements;
}
