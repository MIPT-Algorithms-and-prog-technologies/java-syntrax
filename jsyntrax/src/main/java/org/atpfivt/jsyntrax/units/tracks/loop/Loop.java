package org.atpfivt.jsyntrax.units.tracks.loop;

import org.atpfivt.jsyntrax.exceptions.LoopNotTwoArgsException;
import org.atpfivt.jsyntrax.units.Unit;
import org.atpfivt.jsyntrax.units.tracks.ComplexTrack;
import org.atpfivt.jsyntrax.units.tracks.Track;
import org.atpfivt.jsyntrax.visitors.Visitor;

import java.util.List;

public class Loop extends ComplexTrack {
  public Loop(List<Unit> units) throws LoopNotTwoArgsException {
    super(units);

    if (units.size() != 2) {
      throw new LoopNotTwoArgsException();
    }
  }

  public Track getForwardPart() {
    return (Track) getUnits().get(0);
  }

  public Track getBackwardPart() {
    return (Track) getUnits().get(1);
  }

  public boolean isForwardNull() {
    return getForwardPart() == null;
  }

  public boolean isBackwardNull() {
    return getBackwardPart() == null;
  }

  public void accept(Visitor visitor) {
    visitor.visitLoop(this);
  }
}
