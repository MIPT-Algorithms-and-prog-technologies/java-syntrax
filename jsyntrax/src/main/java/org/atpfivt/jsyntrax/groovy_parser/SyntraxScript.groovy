package org.atpfivt.jsyntrax.groovy_parser

import org.atpfivt.jsyntrax.Configuration
import org.atpfivt.jsyntrax.units.Unit
import org.atpfivt.jsyntrax.units.nodes.Node
import org.atpfivt.jsyntrax.units.nodes.NoneNode
import org.atpfivt.jsyntrax.units.tracks.Choice
import org.atpfivt.jsyntrax.units.tracks.Line
import org.atpfivt.jsyntrax.units.tracks.Track
import org.atpfivt.jsyntrax.units.tracks.loop.Loop
import org.atpfivt.jsyntrax.units.tracks.loop.Toploop
import org.atpfivt.jsyntrax.units.tracks.opt.Opt
import org.atpfivt.jsyntrax.units.tracks.opt.Optx
import org.atpfivt.jsyntrax.units.tracks.stack.Indentstack
import org.atpfivt.jsyntrax.units.tracks.stack.Rightstack
import org.atpfivt.jsyntrax.units.tracks.stack.Stack

class SyntraxScript extends Script {
  Unit node
  String title

  static ArrayList<Unit> unitsToString(Object... units) {
    for (i in 0..<units.length) {
      if (units[i] == null) {
        units[i] = new NoneNode()
      } else if (!(units[i] instanceof Track)) {
        units[i] = new Node(units[i].toString())
      }
    }

    return units as ArrayList<Unit>
  }

  Unit title(String title) {
      this.title = title
      return node
  }

  Line line(Object... units) {
    return node = new Line(unitsToString(units) as ArrayList<Unit>)
  }

  Loop loop(Object... units) {
    return node = new Loop(unitsToString(units) as ArrayList<Unit>)
  }

  Toploop toploop(Object... units) {
    return node = new Toploop(unitsToString(units) as ArrayList<Unit>)
  }

  Choice choice(Object... units) {
    return node = new Choice(unitsToString(units) as ArrayList<Unit>)
  }

  Opt opt(Object... units) {
    return node = new Opt([line(units)] as ArrayList<Unit>)
  }

  Optx optx(Object... units) {
    return node = new Optx([line(units)] as ArrayList<Unit>)
  }

  Stack stack(Object... units) {
    return node = new Stack(unitsToString(units) as ArrayList<Unit>)
  }

  Indentstack indentstack(int indent, Object... units) {
    return node = new Indentstack(indent, unitsToString(units) as ArrayList<Unit>)
  }

  Rightstack rightstack(Object... units) {
    return node = new Rightstack(unitsToString(units) as ArrayList<Unit>)
  }

  Configuration jsyntrax(Track track,
                                Map<String, String> url_map = [:]) {
    return new Configuration(track, url_map)
  }

  @Override
  Object run() {
    return null
  }
}
