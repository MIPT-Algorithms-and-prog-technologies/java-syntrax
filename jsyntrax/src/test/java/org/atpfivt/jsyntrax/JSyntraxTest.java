package org.atpfivt.jsyntrax;

import org.approvaltests.Approvals;
import org.atpfivt.jsyntrax.generators.SVGCanvas;
import org.atpfivt.jsyntrax.generators.SVGCanvasBuilder;
import org.atpfivt.jsyntrax.groovy_parser.SyntraxScript;
import org.atpfivt.jsyntrax.styles.StyleConfig;
import org.atpfivt.jsyntrax.units.tracks.Choice;
import org.atpfivt.jsyntrax.units.tracks.Line;
import org.atpfivt.jsyntrax.units.tracks.loop.Loop;
import org.atpfivt.jsyntrax.units.tracks.loop.Toploop;
import org.atpfivt.jsyntrax.units.tracks.opt.Opt;
import org.atpfivt.jsyntrax.units.tracks.opt.Optx;
import org.atpfivt.jsyntrax.units.tracks.stack.Indentstack;
import org.atpfivt.jsyntrax.units.tracks.stack.Rightstack;
import org.atpfivt.jsyntrax.units.tracks.stack.Stack;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.util.Map;

import static org.atpfivt.jsyntrax.JSyntraxTestUtils.OPTIONS;


public class JSyntraxTest extends SyntraxScript {

    private SVGCanvasBuilder canvasBuilder;

    @BeforeEach
    void setUp() throws IOException {
        StyleConfig s = new StyleConfig(1, false);
        JSyntraxTestUtils.updateStyle(s);
        canvasBuilder = new SVGCanvasBuilder().withStyle(s);
    }

    @Test
    void lineTest() {
        Line l = line('[', "foo", ',', "/bar", ']');
        SVGCanvas c = canvasBuilder.generateSVG(l);
        String result = c.generateSVG();
        Approvals.verify(result, OPTIONS);
    }

    @Test
    void difficultLoopTest() {
        Loop l = loop(line("/forward", "path"), line("backward", "path"));
        SVGCanvas c = canvasBuilder.generateSVG(l);
        String result = c.generateSVG();
        Approvals.verify(result, OPTIONS);
    }

    @Test
    void toploopTest() {
        Toploop l = toploop(line('(', "forward", ')'), line(')', "backward", '('));
        SVGCanvas c = canvasBuilder.generateSVG(l);
        String result = c.generateSVG();
        Approvals.verify(result, OPTIONS);
    }

    @Test
    void choiceTest() {
        Choice choice = choice('A', 'B', 'C');
        SVGCanvas c = canvasBuilder.generateSVG(choice);
        String result = c.generateSVG();
        Approvals.verify(result, OPTIONS);
    }

    @Test
    void optTest() {
        Opt opt = opt('A', 'B', 'C');
        SVGCanvas c = canvasBuilder.generateSVG(opt);
        String result = c.generateSVG();
        Approvals.verify(result, OPTIONS);
    }

    @Test
    void optxTest() {
        Optx optx = optx('A', 'B', 'C');
        SVGCanvas c = canvasBuilder.generateSVG(optx);
        String result = c.generateSVG();
        Approvals.verify(result, OPTIONS);
    }

    @Test
    void stackTest() {
        Stack s = stack(
                line("top", "line"),
                line("bottom", "line")
        );
        SVGCanvas c = canvasBuilder.generateSVG(s);
        String result = c.generateSVG();
        Approvals.verify(result, OPTIONS);
    }


    @Test
    void verticalStackTest() {
        Stack s = stack(
                line('A', 'B'),
                opt("bypass"),
                line("finish")
        );
        SVGCanvas c = canvasBuilder.generateSVG(s);
        String result = c.generateSVG();
        Approvals.verify(result, OPTIONS);
    }

    @Test
    void indentStackTest() {
        Indentstack indentStack = indentstack(3,
                line("top", "line"),
                line("bottom", "line")
        );
        SVGCanvas c = canvasBuilder.generateSVG(indentStack);
        String result = c.generateSVG();
        Approvals.verify(result, OPTIONS);
    }

    @Test
    void rightStackTest() {
        Rightstack rightStack = rightstack(
                line("top", "line", "with", "more", "code"),
                line("bottom", "line")
        );
        SVGCanvas c = canvasBuilder.generateSVG(rightStack);
        String result = c.generateSVG();
        Approvals.verify(result, OPTIONS);
    }

    @Test
    void urlMapTest() {
        Configuration urlMapped = jsyntrax(stack(
                line("attribute", "/(attribute) identifier", "of"),
                line(choice(toploop("/entity_designator", ","), "others", "all"), ':'),
                line("/entity_class", "is", "/expression", ';')
                ),
                Map.of("entity_class", "https://www.google.com/#q=vhdl+entity+class",
                        "(attribute) identifier", "http://en.wikipedia.com/wiki/VHDL"));
        SVGCanvas c = canvasBuilder.generateSVG(urlMapped);
        String result = c.generateSVG();
        Approvals.verify(result, OPTIONS);
    }

    @Test
    void testOptAsLastStackUnit() {
        Configuration opt = jsyntrax(stack(
                line("attribute", "/(attribute) identifier", "of"),
                line(choice(toploop("/entity_designator", ","), "others", "all"), ":"),
                line("/entity_class", "is", "expression", ';'),
                opt("None")
        ));
        SVGCanvas c = canvasBuilder.generateSVG(opt);
        String result = c.generateSVG();
        Approvals.verify(result, OPTIONS);
    }
}