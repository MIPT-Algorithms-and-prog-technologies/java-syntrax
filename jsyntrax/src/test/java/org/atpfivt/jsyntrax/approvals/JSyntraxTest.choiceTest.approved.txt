<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!-- Created by Jyntrax https://github.com/atp-mipt/jsyntrax -->
<svg xmlns="http://www.w3.org/2000/svg"
xmlns:xlink="http://www.w3.org/1999/xlink"
xml:space="preserve"
width="148" height="96" version="1.1">
<style type="text/css">
<![CDATA[
.title_font {fill:#000000; text-anchor:middle;
font-family:PT Sans; font-size:12pt; font-weight:normal; font-style:normal; }
.token_font {fill:#000000; text-anchor:middle;
font-family:PT Sans; font-size:12pt; font-weight:normal; font-style:normal; }
.box_font {fill:#000000; text-anchor:middle;
font-family:PT Sans; font-size:12pt; font-weight:normal; font-style:normal; }
.bubble_font {fill:#000000; text-anchor:middle;
font-family:PT Sans; font-size:12pt; font-weight:normal; font-style:normal; }
.hex_font {fill:#000000; text-anchor:middle;
font-family:PT Sans; font-size:12pt; font-weight:normal; font-style:normal; }
.label {fill: #000; text-anchor:middle; font-size:16pt; font-weight:bold; font-family:Sans;}
.link {fill: #0D47A1;}
.link:hover {fill: #0D47A1; text-decoration:underline;}
.link:visited {fill: #4A148C;}
]]>
</style>
<defs>
<marker id="arrow" markerWidth="5" markerHeight="4" refX="2.5" refY="2" orient="auto" markerUnits="strokeWidth">
<path d="M0,0 L0.5,2 L0,4 L4.5,2 z" fill="#000000" />
</marker>
</defs>
<rect width="100%" height="100%" fill="white"/>
<circle cx="8" cy="16" r="3" stroke="#000000" stroke-width="2" fill="#ffffff"/>
<path d="M72,34 A13,13 0 0,1 72,8 H82 A13,13 0 0,1 82,34 z" fill="#000000" fill-opacity="0.4980392156862745" />
<path d="M69,31 A13,13 0 0,1 69,5 H79 A13,13 0 0,1 79,31 z" stroke="#000000" stroke-width="2" fill="#b3e5fc" fill-opacity="1.0" />
<text class="bubble_font" x="74" y="23">A</text>
<path d="M72,64 A13,13 0 0,1 72,38 H82 A13,13 0 0,1 82,64 z" fill="#000000" fill-opacity="0.4980392156862745" />
<path d="M69,61 A13,13 0 0,1 69,35 H79 A13,13 0 0,1 79,61 z" stroke="#000000" stroke-width="2" fill="#b3e5fc" fill-opacity="1.0" />
<text class="bubble_font" x="74" y="53">B</text>
<path d="M72,94 A13,13 0 0,1 72,68 H81 A13,13 0 0,1 81,94 z" fill="#000000" fill-opacity="0.4980392156862745" />
<path d="M69,91 A13,13 0 0,1 69,65 H78 A13,13 0 0,1 78,91 z" stroke="#000000" stroke-width="2" fill="#b3e5fc" fill-opacity="1.0" />
<text class="bubble_font" x="73" y="83">C</text>
<line x1="28" y1="16" x2="52" y2="16" stroke="#000000" stroke-width="2" marker-end="url(#arrow)" />
<line x1="92" y1="16" x2="121" y2="16" stroke="#000000" stroke-width="2" />
<path d="M37,25 A9,9 0 0,0 28,16" stroke="#000000" stroke-width="2" fill="none"/>
<line x1="46" y1="46" x2="52" y2="46" stroke="#000000" stroke-width="2" marker-end="url(#arrow)" />
<line x1="92" y1="46" x2="102" y2="46" stroke="#000000" stroke-width="2" />
<path d="M37,37 A9,9 0 0,0 46,46" stroke="#000000" stroke-width="2" fill="none"/>
<path d="M102,46 A9,9 0 0,0 111,37" stroke="#000000" stroke-width="2" fill="none"/>
<line x1="46" y1="76" x2="52" y2="76" stroke="#000000" stroke-width="2" marker-end="url(#arrow)" />
<line x1="91" y1="76" x2="102" y2="76" stroke="#000000" stroke-width="2" />
<path d="M37,67 A9,9 0 0,0 46,76" stroke="#000000" stroke-width="2" fill="none"/>
<path d="M102,76 A9,9 0 0,0 111,67" stroke="#000000" stroke-width="2" fill="none"/>
<path d="M120,16 A9,9 0 0,0 111,25" stroke="#000000" stroke-width="2" fill="none"/>
<line x1="37" y1="67" x2="37" y2="25" stroke="#000000" stroke-width="2" />
<line x1="111" y1="67" x2="111" y2="25" stroke="#000000" stroke-width="2" />
<line x1="10" y1="16" x2="24" y2="16" stroke="#000000" stroke-width="2" marker-end="url(#arrow)" />
<circle cx="140" cy="16" r="3" stroke="#000000" stroke-width="2" fill="#ffffff"/>
<line x1="119" y1="16" x2="133" y2="16" stroke="#000000" stroke-width="2" marker-end="url(#arrow)" />
</svg>
